flow/BlockCipher.h

```diff
@@ -0,0 +1,237 @@
+//
+// Supported cache lookups schemes:
+// 1. Lookup cipher based on { encyrptionDomainId, baseCipherKeyId } tuple.
+// 2. Lookup cipher based on BlockCipherEncryptionHeader
+//
+// Client is responsible to handle cache-miss usecase, the corrective operation might vary based on the
+// calling process, for instance: EncryptKeyServer cache-miss shall invoke RPC to external Encryption Key Manager to
+// fetch the required encryption key, however, CPs/SSs cache-miss would result in RPC to EncryptKeyServer to refresh the
+// desired encryption key.
+
+using BlockCipherKeySaltCache = std::unordered_map<BlockCipherRandomSalt, Reference<BlockCipherKey>>;
+
+class BlockCipherKeyItem : public ReferenceCounted<BlockCipherKeyItem>, NonCopyable {
+	Reference<BlockCipherKey> latest;
+	BlockCipherKeySaltCache keyCache;
+
+public:
+	BlockCipherKeyItem(Reference<BlockCipherKey>& cipher) { updateLatest(cipher); }
+
+	Reference<BlockCipherKey> getLatest() { return latest; }
+
+	void updateLatest(Reference<BlockCipherKey> cipher) {
+		latest = cipher;
+		keyCache.emplace(cipher.getPtr()->getSalt(), cipher);
+	}
+	Reference<BlockCipherKey> findCipher(const BlockCipherRandomSalt& salt) {
+		auto itr = keyCache.find(salt);
+		if (itr == keyCache.end()) {
+			return Reference<BlockCipherKey>();
+		}
+		return itr->second;
+	}
+	void reset() {
+		for (auto& keyItr : keyCache) {
+			keyItr.second.getPtr()->reset();
+		}
+	}
+	std::vector<Reference<BlockCipherKey>> getAllCiphers() {
+		std::vector<Reference<BlockCipherKey>> ciphers;
+		for (auto itr : keyCache) {
+			ciphers.emplace_back(itr.second);
+		}
+		return ciphers;
+	}
+};
+
+using BlockCipherKeyIdCacheMap = std::unordered_map<BlockCipherBaseKeyId, Reference<BlockCipherKeyItem>>;
+using BlockCipherDomainCacheMap = std::unordered_map<BlockCipherDomainId, BlockCipherKeyIdCacheMap>;
+
+class BlockCipherKeyCache : NonCopyable {
+	BlockCipherDomainCacheMap domainCacheMap;
+	static uint64_t CIPHER_KEY_CACHE_TTL_NS;
+
+	BlockCipherKeyCache() {}
+
+public:
+	void insertCipherKey(const BlockCipherDomainId& domainId,
+	                     const BlockCipherBaseKeyId& baseCipherId,
+	                     const uint8_t* baseCipher,
+	                     int baseCipherLen);
+	Reference<BlockCipherKey> getLatestCipherKey(const BlockCipherDomainId& domainId,
+	                                             const BlockCipherBaseKeyId& baseKeyId);
+	Reference<BlockCipherKey> getCipherKey(const BlockCipherEncryptHeader& header);
+	std::vector<Reference<BlockCipherKey>> getAllCiphers(const BlockCipherDomainId& domainId,
+	                                                     const BlockCipherBaseKeyId& baseKeyId);
+	static BlockCipherKeyCache& getInstance() {
+		static BlockCipherKeyCache instance;
+		return instance;
+	}
+	// Ensures cached encryption key(s) (plaintext) never gets persisted as part of FDB
+	// process/core dump.
+	static void cleanup() noexcept;
```

Comment:
I had similar thoughts, however, it would be better to make changes (if possible) as a separate patch; for now leveraging the existing framework to ZER-OUT encryption key buffer before generating an FDB process dump.